drivers/imu/adis1654x.c

```diff
@@ -0,0 +1,707 @@
+	default:
+		return -EINVAL;
+	}
+}
+
+/**
+ * @brief Read adis device channel offset in integer.
+ * @param adis      - The adis device.
+ * @param offset    - The read channel offset.
+ * @param chan_type - The channel type to read offset for.
+ * @return 0 in case of success, error code otherwise.
+ */
+static int adis1654x_get_offset(struct adis_dev *adis, int *offset,
+				enum adis_chan_type chan_type)
+{
+	switch(chan_type) {
+	case ADIS_TEMP_CHAN:
+		*offset = adis1654x_temp_offset[ADIS1654X_ID_NO_OFFSET(adis->dev_id)];
+		return 0;
+	default:
+		return -EINVAL;
+	}
+}
+
+/**
+ * @brief CRC32 computation algorithm for adis1654x burst access.
+ * @param len  - size of buff in bytes.
+ * @param buff - data buffer to check.
+ * @return computed crc.
+ */
+static uint32_t adis1654x_crc32_computation(uint32_t len, const uint8_t *buff)
+{
+	uint32_t crc = ADIS1654X_CRC32_SEED;
+	unsigned int j;
+	uint32_t poly = 0xEDB88320;
+
+	while (len > 0) {
+		crc ^= *buff;
+
+		for (j = 8; j > 0; j--) {
+			if (crc & 1)
+				crc = crc >> 1 ^ poly;
+			else
+				crc = crc >> 1;
+		}
+		buff++;
+		len--;
+	}
+
+	crc = crc ^ 0xFFFFFFFFU;
+
+	return crc;
+}
+
+/**
+ * @brief Read burst data.
+ * @param adis      - The adis device.
+ * @param data      - The burst read data structure to be populated.
+ * @param burst32   - True if 32-bit data is requested for accel
+ *		      and gyro (or delta angle and delta velocity)
+ *		      measurements, false if 16-bit data is requested.
+ * @param burst_sel - 0 if accel and gyro data is requested, 1
+ *		      if delta angle and delta velocity is requested.
+ * @param fifo_pop  - In case FIFO is present, will pop the fifo if
+ * 		      true. Unused if FIFO is not present.
+ * @param crc_check - If true CRC will be checked, if false check will be skipped.
+ * @return 0 in case of success, error code otherwise.
+ * -EAGAIN in case the request has to be sent again due to data being unavailable
+ * at the time of the request.
+ */
+int adis1654x_read_burst_data(struct adis_dev *adis,
+			      struct adis_burst_data *data,
+			      bool burst32, uint8_t burst_sel, bool fifo_pop, bool crc_check)
+{
+	int ret;
+	uint8_t msg_size = 46;
+	uint8_t burst_id = 0xA5;
+	uint32_t computed_crc32, recv_crc32;
+
+	if (!burst32)
+		return -EINVAL;
+
+	if (adis->info->flags & ADIS_HAS_BURST32) {
+		if (adis->burst_sel != burst_sel) {
+			ret = adis_write_burst_sel(adis, burst_sel);
+			if (ret)
+				return ret;
+			/* If burst select has changed, wait for the next reading
+			request to actually read the data, because the according data will be available
+			only after the next data ready impulse. */
+			return -EAGAIN;
+		}
+	}
+
+	if (adis->current_page != 0) {
+		adis->tx[0] = ADIS_WRITE_REG(ADIS_REG_PAGE_ID);
+		adis->tx[1] = 0;
+		ret = no_os_spi_write_and_read(adis->spi_desc, adis->tx, 2);
+		if (ret)
+			return ret;
+		adis->current_page = 0;
+	}
+
+	if (burst_sel)
+		burst_id = 0xC3;
+
+	uint8_t buffer[msg_size], offset;
```

Comment:
yeah like Toni said, let's move these local variables at the top, valid for the whole file